File: aufgabe_2_caching/php/WeatherProxy.php
<?php

interface WeatherInterface {
    public function getWeatherData(string $city): string;
}

// Der echte Service (simuliert eine langsame API)
class WeatherService implements WeatherInterface {
    public function getWeatherData(string $city): string {
        echo "[API] Rufe Daten vom Wetter-Server ab für: {$city}...\n";
        sleep(2); // Simuliert 2 Sekunden Netzwerk-Verzögerung
        return "22°C, Sonnig";
    }
}

class WeatherProxy implements WeatherInterface {
    private WeatherService $weatherService;
    private array $cache = []; // Hier werden die Ergebnisse gespeichert

    public function __construct() {
        $this->weatherService = new WeatherService();
    }

    public function getWeatherData(string $city): string {
        // TODO: Prüfe, ob die $city bereits als Key im Array $this->cache existiert
        
        // Wenn JA: Gib den gespeicherten Wert direkt zurück und schreibe:
        // echo "[Proxy] Daten aus dem Cache geladen für: {$city}\n";

        // Wenn NEIN: Rufe den echten Service auf, speichere das Ergebnis im Cache
        // und gib den Wert zurück.
        
        return ""; // Temporärer Rückgabewert
    }
}

// --- TEST BEREICH ---
// $proxy = new WeatherProxy();
// echo $proxy->getWeatherData("Bonn") . "\n"; // Erste Abfrage: Dauert 2 Sekunden (API-Call)
// echo $proxy->getWeatherData("Bonn") . "\n"; // Zweite Abfrage: Sofort da (aus dem Cache!)

File: aufgabe_2_caching/python/weather_proxy.py
import time
from abc import ABC, abstractmethod

class WeatherInterface(ABC):
    @abstractmethod
    def get_weather_data(self, city: str) -> str:
        pass

class WeatherService(WeatherInterface):
    def get_weather_data(self, city: str) -> str:
        print(f"[API] Rufe Daten vom Wetter-Server ab für: {city}...")
        time.sleep(2) # Simuliert Verzögerung
        return "22°C, Sonnig"

class WeatherProxy(WeatherInterface):
    def __init__(self):
        self.weather_service = WeatherService()
        self.cache = {} # Dictionary für das Caching

    def get_weather_data(self, city: str) -> str:
        # TODO: Prüfe, ob city bereits im Dictionary self.cache existiert
        
        # Wenn JA: Printe "[Proxy] Daten aus dem Cache geladen" 
        # und gib den gespeicherten Wert zurück.

        # Wenn NEIN: Rufe self.weather_service.get_weather_data(city) auf,
        # speichere das Ergebnis im Cache und gib es zurück.
        pass

# --- TEST BEREICH ---
# proxy = WeatherProxy()
# print(proxy.get_weather_data("Bonn")) # Erste Abfrage: Dauert 2 Sekunden
# print(proxy.get_weather_data("Bonn")) # Zweite Abfrage: Sofort da!
